### DISCUSSION TITLE
Auto-Reset SoC to 100% when upper voltage limit is reached

### DISCUSSION THREAD

**Discussion by transistorgit** [Louisvdw/dbus-serialbattery#657] [Ideas]
Hi,

I have 2 Daly Smart BMS and a SmartShunt. Obviously the dalys are quite bad at current measurements. Altough I could try to calibrate them better, I would like to have the same behaviour like in the SoC measurement of the smartshunt. It resets the SOC to 100% every time it reaches the upper voltage level (I'm not sure how it is labeled there, I think absorbtion voltage).

I can implement it in the daly driver, but think it would be probably an interesting feature for other bms as well? What do the users here think? Please comment.

brgds

**Comment by RayTIW** [comment]
Hi,

I use the same setup (2 x Daily Smart BMS and Victron SmartShunt)! I would like to see this feature, too!

BR

Ray

**Reply by josmeijer** [reply]
A remark I must make: if you assume that max_charging voltage = SOC100% you do not account with decaying cells over time., so you might be using a 230 AH battery that has only 130 AH left and might be unpleasantly surprised by an empty battery when least convenient.
I admit this is "Lead-Acid Science" but yet I guess it is something to consider.

**Comment by drtinaz** [comment]
This would definitely be a welcomed feature.

**Reply by Gr33n93** [reply]
Hello,

Also, is the warning message from any Daly BMS at below 20% taken into account?

I get every time the message that the charge level is low and then see on the display only the warning message.

Is it possible to configure this message as disengageable?
Since this comes with each discharge daily below 20%.

I myself use 3 Daly Smart BMS and would find the function also important.

**Reply by drtinaz** [reply]
You might take a look at node red and mqtt. I have instituted a simple flow in node red on my system using mqtt that automatically syncs the bms SOC to the smart shunt SOC when they differ more than a set amount.

**Reply by transistorgit** [reply]
You can set the warning level by configuring it in the bms itself, using the daly pc master software. The driver just shows what the bms sends.

**Reply by Gr33n93** [reply]
Thanks for the feedback.
I had only searched for the function in the app and therefore had not rounded it.

**Reply by RichieS67** [reply]
@drtinaz:
I've been looking for such a flow like this for a long time! I can overwrite the Daly BMS SoC with the Victron Shunt SoC, but a few seconds later it is back to the original incorrect value. Could you please send me your flow? Thank you - Richie

**Comment by mr-manuel** [comment]
The JKBMS for example resets the SoC to 95% when all cell voltages are greater or equal to `OVP` (setting in the JKBMS app). So no need for JKBMS at the moment. There is also no possibility to reset the SoC with a command.

**Comment by rperpe** [comment]
JK BMS user here with same issue; SoC measurement isn't configurable at all and coloumb metering is unaccurate at low currents. Consequence: SoC is getting lower and lower.

As a solution I thought about using **another SoC source than the BMS** (e.g. the SmartShunt). Is this somehow feasible in dbus-serialbattery?

**Reply by mr-manuel** [reply]
No but with [dbus-mqtt-battery](https://github.com/mr-manuel/venus-os_dbus-mqtt-battery) and Node-RED (integrated in Venus OS large) you can mix all values you want

**Comment by cokyc** [comment]
I use a daly bms and a victron smartshunt.
I would like to sync smartshunt soc with the one read by serialdriver and tried to set soc with node-red.
I wrote, for example {"value": 87.5} to topic W/XXXXXXXX/battery/1/Settings/ResetSoc but do not work.
This topic seems read-only.

**Reply by drtinaz** [reply]
Something must not be set properly because it is a writeable setting. I do exactly what you are trying to do and I have no issues keeping things in sync.

**Reply by transistorgit** [reply]
please check the log. I implemented it while using the UART interface and had no problems. Now I switched to RS485 and one of my dalys (I think the one with older firmware) won't accept the setting. You would find a log message like "write soc .... failed"

what interface are you using?

**Comment by cokyc** [comment]
Hi,
yesterday I was using RS485 and It was not working in any way.
Today I switched to TTL and I think I am doing something wrong (but still I don't know what) in node-red because,
connecting to venus-os with MQTT explorer I can send values to soc correctly but not with node-red.
And with node-red seems I can't send anything to any topic.
I can see the node mqtt-out connected to the broker.

**Reply by mr-manuel** [reply]
Please share the payload you are sending via MQTT Explorer, the Node-RED Flow and the Payload you see in MQTT Explorer that is send from Node-RED. Maybe you are sending the number as string.

**Reply by Markus1301** [reply]
Hi,
I want to write the SOC to my Daly BMS. I'm confused about this threat. Do I need your dbus mqqt battery to do this? I saw the pictures of it. Is it a replacement for dbus serial battery or an add on?

**Comment by cokyc** [comment]
The payload is:
{"value":91}
the topic is:
W/e45f01de751c/battery/1/Settings/ResetSoc

![immagine](https://github.com/Louisvdw/dbus-serialbattery/assets/61275051/8fe6104a-f578-4614-872e-2cc47325cc55)

**Reply by Markus1301** [reply]
Hallo,
I need this, too. It would be very nice, if you could write some more for a beginner, what I have to do here?
Belongs payload and Topic to the "inject".  Is payload jason because of the { }
What should I do with the mqtt out. What should I type in "Server"? To I have to install something more, like this dbus-mqtt-battery?

As I said: I little help would be very nice.

**Reply by cokyc** [reply]
Hello Markus,
As in the picture above, you have to add an injection node where you can delete the timestamp value and keep only the payload value.
Format it as JSON. Set it as {"value":xx} and override the xx with your value. Link the inject node to an MQTT out node.
Set it as in the picture, but change the "e45f01de751c" part with your value (you can see it with mqttexplorer connected to venus-os).
Of course, you have to change the inject node with one where you read the soc from. (for example from a smartshunt or as mr-manuel suggested you can use his [dbus-mqtt-battery](https://github.com/mr-manuel/venus-os_dbus-mqtt-battery) driver

**Reply by Markus1301** [reply]
Thank you for your explanation. I don't have a smart shunt, I use my Daly BMS with dbus-serial-battery at the moment. So I looked at dbus-mqtt-battery. There are pictures which look like dbus-serialbattery in VenusOS. Is dbus-mqtt-battary an addon to dbus-serial-battery or is it a replacement?

**Comment by cokyc** [comment]
ok. Solved.
I had a mistake and I am not seeing it.
I was sending the payload as topic in the injection node.

My bad.

**Comment by transistorgit** [comment]
Just a small status update: 
I implemented the auto-reset-soc on my system, but need to improve my debug logging a bit to make sure it works correctly. 

It sets the SOC to 100% when battery status switches to float. 
Dbus-spy: Info/ChargeMode

**Reply by Markus1301** [reply]
Hi,
first of all: great work! Thanks a lot. This is, what I need.

pro: When my battery is full (55,2V), I can set the SOC to 90% per hand in VenusOS. Great. This helps a lot. Before this I had to unplug the BMS from the raspberry, use a Notebook with the Daly Software to reset the SOC.

But I still have a problem: It doesn't work automatic. Maybe it is because my system never goes to float. When the battery is full (55,2V) it stops charging and there is "idle". In my config.ini is only:
[DEFAULT]

; If you want to add custom values/settings, then check the values/settings you want to change in "config.default.ini"
; and insert them below to persist future driver updates.

; Example (remove the semicolon ";" to uncomment and activate the value/setting):
MAX_BATTERY_CHARGE_CURRENT = 70.0
MAX_BATTERY_DISCHARGE_CURRENT = 180.0

; Daly settings
; Battery capacity (amps) if the BMS does not support reading it 
BATTERY_CAPACITY = 280
; Invert Battery Current. Default non-inverted. Set to -1 to invert
INVERT_CURRENT_MEASUREMENT = -1

I have 1) latest installed. Should I go to nightly, or something else?

Thanks for any help.
Markus

**Reply by mr-manuel** [reply]
You have to use nightly for the auto SoC reset, since it wasn't released in stable yet.

**Reply by Markus1301** [reply]
Hi, thanks for your reply. I changed to nightly and there to dev. Now VenusOS says 55,1V. The battery is not charging any more (there are lot's of PV power at the moment). In VRM portal the Multiplus II says: IDLE.  On the Multiplus itself the LED absorbtion is on. I never saw the "float" LED on. And the battery is 83,7%, not 90%. So the automatic reset of the SOC doesn't work in my case.

My config.in now is:
[DEFAULT]

; If you want to add custom values/settings, then check the values/settings you want to change in "config.default.ini"
; and insert them below to persist future driver updates.

; Example (remove the semicolon ";" to uncomment and activate the value/setting):
MAX_BATTERY_CHARGE_CURRENT = 70.0
MAX_BATTERY_DISCHARGE_CURRENT = 180.0

; Daly settings
; Battery capacity (amps) if the BMS does not support reading it 
BATTERY_CAPACITY = 280
; Invert Battery Current. Default non-inverted. Set to -1 to invert
INVERT_CURRENT_MEASUREMENT = -1

; -- CVL reset based on SoC option (step mode & linear mode)                                                        
; Specify how long the max voltage should be kept                                                                   
;     Step mode: If reached then switch to float voltage                                                            
;     Linear mode: If cells are balanced keep max voltage for further MAX_VOLTAGE_TIME_SEC seconds                  
MAX_VOLTAGE_TIME_SEC = 30                                                                                          
; Specify SoC where CVL limit is reset to max voltage                                                               
;     Step mode: If SoC gets below                                                                                  
;     Linear mode: If cells are unbalanced or if SoC gets below                                                     
SOC_LEVEL_TO_RESET_VOLTAGE_LIMIT = 90   

Edit 2: A few hours later. The SOC ist now 99% (53,9V). How can this be. In the config.ini (see above) is SOC_LEVEL_TO_RESET_VOLTAGE_LIMIT = 90. I thought it resets to 90%. Is it the wrong parametet?

Markus

**Reply by mr-manuel** [reply]
The MultiPlus Float/Absorption LEDs do not correspond to the driver state but shows only the MutliPlus algorithm, even if it's the active one.

**Reply by Markus1301** [reply]
Thanks for your feedback. One question left: I thought SOC_LEVEL_TO_RESET_VOLTAGE_LIMIT = 90 in the config.ini File will set the SOC to 90% when 55,2V are reached. It goes to 100%. Is this a bug or did I misunderstand something, again?

**Reply by mr-manuel** [reply]
You misunderstood it.

https://github.com/Louisvdw/dbus-serialbattery/blob/bc267e16138ac8ee19cc2869d067c7b2f233118a/etc/dbus-serialbattery/config.default.ini#L79-L87

If the SoC gets below this value, the absorption voltage is allowed again. This has nothing to do with the SOC reset.

**Comment by cokyc** [comment]
why not replicate the behaviour of smartshunt, reset soc to 100% when voltage is  xx volt for n minutes.
For example 3.45x16 for 3 minutes.
Or we can put two parameters in config.ini AutoResetSocVoltage and AutoResetSocDuration

It work very well for smartshunt.

**Reply by transistorgit** [reply]
Actually its the same, as the charging modes are calculated out of the voltages

**Comment by transistorgit** [comment]
Hi,
I created a pull request for @mr-manuel 's dev branch and he already merged it. So if somebody wants to try...

**Reply by mr-manuel** [reply]
It was merged to the `dev` branch of louisvdw and can be installed with the install script.

**Comment by cokyc** [comment]
Installed yesterday and today have tested it.
**It worked** but I am not so sure about the timing.
My smartshunt is set to go to 100% after 4 minutes the voltage is over  55.2.
I have MAX_VOLTAGE_TIME_SEC = 200 and MAX_CELL_VOLTAGE   = 3.46 in config.ini file.

Smart shunt soc was reset to 100% after 4 min of voltage over 55.2
BMS after about 16 min, so seems the setting in config.ini is ignored and the default 900 sec in config.default.ini has been considered


Here is a graph.
![immagine](https://github.com/Louisvdw/dbus-serialbattery/assets/61275051/790ceb0c-546a-4e20-a471-826f91cdf96a)

**Reply by mr-manuel** [reply]
You can check under parameters in the remote console/GUI what's going on. Probably you have linear mode turned on. In this case float is triggered only, if batteries are balanced first. Just check the page at the given time.

![grafik](https://github.com/Louisvdw/dbus-serialbattery/assets/12592765/9f5119ca-a737-4b99-bc7c-48ec45f84c10)

**Comment by cokyc** [comment]
yes I have linear mode on.
Thank you for the explanation. 
I will check in the next days.

**Comment by cokyc** [comment]
I installed the last dev version.
Now I am able to reset soc also in rs485.
But not all the time the soc reset to 100%, because often when the voltage is about 54.6/54.8 the multiplus (I don't know why) goes to absorption and set the charge current about to zero.
Consequently, the max voltage (55.2V) is never reached.

**Comment by josmeijer** [comment]
The voltage to switch from absorption to float can be freely adjusted in
the Victron configuration software.

Op ma 3 jul. 2023 17:50 schreef cokyc ***@***.***>:

**Reply by cokyc** [reply]
I already set absorption and float in ve config the same as serialbus driver in config.ini:
absorption = 55.2V (3.45 x 16)
float = 54 V (3.375x16)
